Remove the C Kconfig tools

ulfalizer opened the pull request:
This patchset removes the C Kconfig tools, switching over to the new `menuconfig.py` menuconfig implementation. It also updates the documentation.

Please scream if you spot something you use being removed.

You can already try out the new `menuconfig` interface by running e.g. `ninja pymenuconfig`.

See the commit messages for more information.

ulfalizer committed:
doc: Update for menuconfig.py

Update the Application Development Primer to describe the new menuconfig
configuration interface:

 - Update the section explaining how the configuration is navigated,
   changed, and saved

 - Add a section explaining how to jump directly to a symbol in the
   configuration interface

 - Replace mconf screenshots with menuconfig.py screenshots. Remove some
   screenshots that are no longer used.

 - Remove the section explaining how to load arbitrary .config files
   from within the configuration interface. That feature hasn't been
   implemented yet in menuconfig.py.

 - Do various minor language cleanup

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
kconfig: Remove targets specific to the C implementation

They will no longer be available once the C Kconfig implementation is
removed.

oldconfig and allno/yesconfig implementations are available for
kconfiglib and could be added later if needed. savedefconfig (minimal
configuration generation) is available from the menuconfig.py
configuration interface.

cmake/usage/kconfig-usage.cmake becomes kinda pointless after this
change, so merge it into cmake/usage/usage.cmake.

Remove the kconfig_target, COMMAND_FOR_*, and COMMAND_RUNS_ON_WIN_*
CMake variables, as there's just the 'menuconfig' target now.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
scripts: kconfig: Switch 'menuconfig' over to menuconfig.py

Make 'make/ninja menuconfig' run menuconfig.py instead of mconf from the
C Kconfig tools. Get rid of the 'pymenuconfig' target.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
kconfig: Remove the C Kconfig implementation

Remove the C Kconfig tools and various scripts associated with them.

scripts/kconfig/diffconfig is popular, so keep it.

I don't know whether anyone is using scripts/kconfig/config. Remove it
and see if anyone screams.

scripts/kconfig/streamline_config.pl deals with modules ('m' values) and
can safely be removed. Zephyr's Kconfig files do not use modules.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
kconfig: Mention that checkconfig.py lacks Kconfiglib 2 support

This script has not been updated for Kconfiglib 2. Add a comment to it
that mentions it.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer commented:
CC @mbolivar @carlescufi

carlescufi reviewed (changes requested):
Blocking this until we agree on how to present `windows-curses` as a requirement

ulfalizer commented on doc/getting_started/installation_win.rst:
```diff
@@ -85,7 +85,12 @@ packages from their respective websites.
 
       cd %userprofile%\zephyr
-      pip3 install --user -r scripts/requirements.txt
+      pip3 install --user -r scripts/requirements.txt windows-curses
```
Looks like it's possible: https://stackoverflow.com/a/11704396/4577158

I'll change it.

carlescufi replied:
Great, this is the better solution IMHO.

mbolivar replied:
Yeah, this is cool

carlescufi replied:
So what's the conclusion here? I vote for having scripts/requirements-windows.txt` instead of having to type `windows-curses` but I don't mind if we keep `pip3 install --user -r scripts/requirements.txt windows-curses`

mbolivar replied:
+1, though I think it should be `windows-requirements.txt` for parallelism with the existing `py2-requirements.txt` (i.e. keep the pattern that our requirements.txt files always end in `requirements.txt`, instead of being `requirements-xxx.txt` sometimes but not always).

carlescufi commented on doc/application/application.rst:
```diff
@@ -935,155 +935,91 @@ The example below shows a comment line and an override setting
-   a) Using CMake, create a build directory (:file:`<home>/app/build`) from
-      your application directory (:file:`<home>/app`).
+#. Create a build directory :file:`<home>/app/build` inside your application
+   directory and generate a build system inside it with CMake, as follows:
```
replace "generate a build system" with "generate all required build files"

ulfalizer replied:
What about "generate build files"?

carlescufi replied:
sounds good

carlescufi commented on doc/application/application.rst:
```diff
@@ -935,155 +935,91 @@ The example below shows a comment line and an override setting
+     .. note::
 
-   * Use the arrow keys to navigate within any menu or list.
+        Common `Vim <https://www.vim.org>`_ navigation commands are
```
replace "commands" with "key bindings"

ulfalizer replied:
Will do

carlescufi commented on doc/application/application.rst:
```diff
@@ -935,155 +935,91 @@ The example below shows a comment line and an override setting
-      change the file name to :file:`.config`. To keep your original
-      :file:`.config`, rename it to something other than :file:`.config`.
-
-   Kernel configuration files, such as the :file:`.config` file, are saved
```
Why did you remove this paragraph? Seems to still apply and there's valuable info there.

ulfalizer replied:
It wasn't removed, just dedented. See below. :)

ulfalizer replied:
Ah, sorry, I see which paragraph you mean now.

I wonder how helpful it is in practice to explain how to list files starting with `.` here. The description is *nix-specific as well, as those files are always shown on Windows.

Maybe a short note along the lines of `Note that 'ls' does not list files starting with '.' by default. You can pass the '-a' flag to see them.` would be enough, if we keep it.

carlescufi replied:
yes, maybe something like "The Zephyr configuration file generated during CMake configuration is stored in `zephyr/.config`, and is not visible by default on Linux and macOS. Use `ls -a` to be able to see it"

carlescufi commented on cmake/kconfig.cmake:
```diff
@@ -18,26 +18,14 @@ set(ENV{KERNELVERSION}      ${PROJECT_VERSION})
-
-set(COMMAND_FOR_config     ${KCONFIG_CONF} --oldaskconfig ${KCONFIG_ROOT})
-set(COMMAND_FOR_gconfig    gconf                          ${KCONFIG_ROOT})
+set(kconfig_target_list menuconfig)
```
I wonder if we need a target list anymore? Could we simplify this? Maybe a follow-up PR?

carlescufi commented on scripts/kconfig/checkconfig.py:
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env python3
 
+# NOTE: This script has not been updated for Kconfiglib 2 and will not run
```
so why do we keep this script here at all?

ulfalizer replied:
Thought it might be more likely that someone might update it if it's kept. I have no strong opinions here though.

carlescufi replied:
@nashif any comments on this script?

carlescufi commented:
>diffconfig is actually a nice script that we should keep IMO

I agree. Does it need changes to work with the future `rsource` and `gsource` directives?

nashif commented:
> I agree. Does it need changes to work with the future rsource and gsource directives?

no, this does not use the backend, it just does python based diffing on two configs

carlescufi commented:
@ulfalizer During the TSC call today I mentioned this PR and there was one objection to get rid of the C tools: search support. Would it be possible to add this to `pymenuconfig` before we merge this PR in order to make sure we do not disturb the workflow of the different developers using Zephyr?

mbolivar commented on doc/application/application.rst:
```diff
@@ -935,155 +935,91 @@ The example below shows a comment line and an override setting
-   If you want to permanently alter the configuration you should set
-   the new value in a :file:`.conf` file, as described above in
-   :ref:`application_set_conf`.
+The configuration can also be changed by editing :file:`zephyr/.config` by
```
Perhaps:

```
by editing :file:`zephyr/.config` in the application build directory
```

mbolivar reviewed (commented):
@ulfalizer BTW, it was my concern at the TSC meeting that before we rip out the existing C implementations, we should at least have a search feature, if not both search and jump to definition.

Without that, this is a feature regression for users of the C tools, and it's no longer a no-op to merge.

ulfalizer commented:
Morning

Alright, I'll add back `diffconfig`. I removed everything in the PR just to see what people are attached to. :)

Adding search/jump-to before integration is fine by me. I'll start working on it.

ulfalizer commented:
Show-all mode has been implemented:

![ss9](https://user-images.githubusercontent.com/154768/39554656-6f71f39c-4e74-11e8-85fc-9dfdc9f2913c.png)

For the search/jump-to window, I'm thinking of having an edit box at the top, with a list of matching symbol names appearing below it as you type. You could then step into the list with the up/down arrow keys, and jump to the selected symbol with `Enter`.

Show-all mode is needed since the symbol you want to jump to might not be visible. It will be enabled automatically in that case.

ulfalizer commented:
Fixed some other minor stuff:

  - The save-and-quit dialog is now skipped if nothing has been changed

  - Bool/tristate symbol values can now be set by pressing `N`/`M`/`Y`, like in `make menuconfig`

  - The menu path at the top now scrolls to the right if needed to show the title of the current menu

Search should be next.

ulfalizer commented:
Search and jump-to is coming along. Still some missing stuff to add and code to clean up.

Screenshot (interface description at bottom):

![skarmbild fran 2018-05-04 07-58-21](https://user-images.githubusercontent.com/154768/39614627-863f862c-4f71-11e8-8640-853ad6266e85.png)

I'm designing it so that you can (ideally) just type part of a symbol name and press Enter to jump to it.

Here's where you end up for the example above:

![skarmbild fran 2018-05-04 07-58-45](https://user-images.githubusercontent.com/154768/39614690-0395037c-4f72-11e8-83a0-0caa1a0afdc5.png)

The symbol wasn't currently visible, so show-all mode was turned on automatically. Red means currently invisible.

ulfalizer committed:
scripts: kconfig: Add incremental search to menuconfig

Pressing [/] brings up a dialog with an edit box where a regex can be
entered. The list of matching symbols is always shown below it.
Selecting a symbol and pressing [Enter] jumps directly to it in the menu
tree. If the symbol is invisible, show-all mode is turned on
automatically.

This commit also includes a bunch of more-or-less unrelated changes from
poking around with the code:

  - Some redundant styles were merged. Probably wouldn't want to have a
    different style for each separator line, for example...

  - [ESC] in the top menu now works like [Q]

  - Returning to a parent menu now makes sure that the selected row is
    visible, even if the terminal was shrunk between entering the child
    menu and leaving it.

  - A _max_scroll() helper was factored out to reduce code duplication.
    It takes a list of items and a window in which the list is
    displayed, with one row per item, and returns the minimum scroll
    value that will make the final item visible.

  - The save dialog now pops up a message to confirm that the save was
    successful.

  - Lots of minor code nits all over (renamings, etc.)

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
scripts: install windows-curses package on Windows

The new menuconfig implementation needs it when run on Windows.

Use an environment marker to only install windows-curses on Windows. See
https://www.python.org/dev/peps/pep-0508/#id23.

From some googling, sys_platform might be more widely supported than
platform_system, so use that.

Suggested-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer commented:
Changes in new PR:

  - Added search and jump-to to the menuconfig implementation. It also supports regular expressions now:

    ![ss11](https://user-images.githubusercontent.com/154768/39704081-8da98106-520a-11e8-98ab-bc22a1c6c3cd.png)

    ![ss12](https://user-images.githubusercontent.com/154768/39704114-a69d9878-520a-11e8-978b-76bcc637eec3.png)

  - Put the Windows requirements (with the extra `windows-curses` package) into a new `scripts/windows-requirements.txt` file.

    There's now a separate `scripts/common-requirements.txt` that lists packages needed on all systems. It's sourced by the system-specific requirements files.

    `scripts/requirements.txt` could be split into `scripts/linux-requirements.txt` and `scripts/mac-requirements.txt` as well, but I kept it for now. There are no Linux- or Mac-specific requirements yet.

  - Added back `diffconfig`

  - Added back note explaining how to list files starting with `.`

  - Fixed the wording issues in the documentation that were pointed out

carlescufi reviewed (approved):
I am happy with this, works fine on Windows. There are tweaks to the UI I would like to get in, but those are non-critical. All the comments I made have been addressed.

carlescufi commented:
Tested on macOS as well with good results, +1

SebastianBoe commented:
@ulfalizer : There are some remnants of mconf in the build system that should be removed.

```
cmake/host-tools.cmake
32:  KCONFIG_MCONF
```

SebastianBoe commented on cmake/kconfig.cmake:
```diff
@@ -18,28 +18,14 @@ set(ENV{KERNELVERSION}      ${PROJECT_VERSION})
-set(COMMAND_RUNS_ON_WIN_pymenuconfig 1)
+set(kconfig_target_list menuconfig)
+set(COMMAND_FOR_menuconfig ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/kconfig/menuconfig.py ${KCONFIG_ROOT})
+set(COMMAND_RUNS_ON_WIN_menuconfig 1)
```
We don't need this construct (RUNS_ON_WIN) now that everything runs on all platforms.

carlescufi replied:
We don't need it in `kconfig.cmake` but it still makes sense in other places, like for example QEMU?

ulfalizer replied:
Yeah... `kconfig_target_list` is a bit silly now too.

SebastianBoe replied:
Correct, but this construct is not used in QEMU. So it is safe to remove it.

carlescufi replied:
Agreed, let's get rid of it

ulfalizer replied:
Got rid of it.

SebastianBoe commented:
This must be blocked until @dbkinder has had time to review the doc changes.

I didn't like some of the new phrasing, but was able to find better language.

carlescufi commented:
>This must be blocked until @dbkinder has had time to review the doc changes.

Fine with me, I thought he had already done a pass at it but I now see it was only me.

SebastianBoe commented on scripts/CMakeLists.txt:
```diff
@@ -1,11 +0,0 @@
-cmake_minimum_required(VERSION 3.6)
```
Now that conf is no longer a dependency we need to update the mac documentation that instructs users to build conf.

carlescufi replied:
Nice catch. @ulfalizer can you get rid of the paragraph beginning with "Build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in  https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/installation_mac.rst

carlescufi replied:
Same for the paragraph beginning with "Within the MSYS console, build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/installation_win.rst

carlescufi replied:
And same with the paragraph beginning with "On UNIX platforms, Build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/getting_started.rst. Also remove "Now that the host tools are installed,"

ulfalizer replied:
Fixed

carlescufi commented on doc/getting_started/installation_win.rst:
```diff
@@ -85,7 +85,7 @@ packages from their respective websites.
 
       cd %userprofile%\zephyr
-      pip3 install --user -r scripts/requirements.txt
+      pip3 install --user -r scripts/windows-requirements.txt
```
I think there's no `if`, but rather `windows-curses; sys_platform == 'win32'`. See here:
https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers

I'm not against trying this out actually, it does keep things clean in terms of requirements files.

carlescufi replied:
I am pretty sure this is safe actually. Python assigns `win32` to `sys.platform` on all Windows versions no matter the bitness of Python itself either, so I would go with this.

carlescufi replied:
I tested this now on both Windows and Linux and can confirm it works fine. My vote is to go back to a single `requirements.txt` and use `windows-curses; sys_platform == 'win32'`

ulfalizer replied:
Nice find.

I changed it. Looks like `sym_platform` might be reasonably safe.

dbkinder commented on doc/application/application.rst:
```diff
@@ -935,155 +935,121 @@ The example below shows a comment line and an override setting
+        symbol to the corresponding value.
 
-#. Press :kbd:`/` to bring up a search menu to look for a particular option.
+#. To view information about a symbol, select it and press :kbd:`?`.
```
The previous key explanations were "active", so let's make this one as well:

    #. Press :kbd:`?` to view information about the selected symbol.

ulfalizer replied:
Hmm... that makes it sound like something the user must do, since it's currently "step 4".

Think that's the real problem. I'll see if I can reorganize it.

dbkinder commented on doc/application/application.rst:
```diff
@@ -935,155 +935,121 @@ The example below shows a comment line and an override setting
 
-#. After configuring the kernel options for your application, tab to
-   :guilabel:`< Save >` and press :kbd:`Enter`.
+   Pressing :kbd:`ESC` or :kbd:`Q` returns from the information display to the
```
Here too:

       Press :kbd:`ESC` or kbd:`Q` to return from the information display to the
       menu display.

dbkinder commented on doc/application/application.rst:
```diff
@@ -935,155 +935,121 @@ The example below shows a comment line and an override setting
-   The following dialog opens, displaying the file name the configuration
-   was saved to.
+      Also note that filenames starting with ``.`` are not listed by ``ls`` by
+      default on Linux and Mac OS X. Use the ``-a`` flag to see them.
```
The name now is simply ``macOS`` instead of ``Mac OS X``.

dbkinder commented on doc/application/application.rst:
```diff
@@ -935,155 +935,121 @@ The example below shows a comment line and an override setting
 
-#. To load any saved kernel configuration file, tab to :guilabel:`< Load >` and
-   press :kbd:`Enter`.
+If you jump to a symbol that isn't currently visible -- e.g. due to having
```
better is to write this as:

    .. visible (e.g., due to having unsatisfied dependencies) then ...

ulfalizer replied:
Got complaints when I added a comma after 'e.g.' in a different project. Maybe I should switch back. ;)

dbkinder replied:
Ha. ;)  Most (but not all) writing style guides (e.g., Chicago Manual of Style, The Columbia Guide to Standard American English) recommend using a comma.  If you had used the phrase "for example" a comma would be expected.

dbkinder commented on doc/application/application.rst:
```diff
@@ -935,155 +935,121 @@ The example below shows a comment line and an override setting
-   command line.
+To figure out why a symbol you jumped to isn't visible, inspect its
+dependencies by pressing :kbd:`?`. If you discover that the symbol depends on
+another symbol that isn't enabled, you can jump to that symbol in turn to see
```
add some commas:

    jump to that symbol, in turn, to see

dbkinder reviewed (changes requested):
Looks pretty good.  A few suggestions.

ulfalizer commented:
@dbkinder 
Fixed.

Reorganized things slightly so that `?` isn't listed as a separate step. Check if it looks alright.

dbkinder reviewed (approved):
Works for me, thanks!
